pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,171 @@
+			// Check if we have converted this ServiceRole before.
+			permissions := make([]*policyproto.Permission, 0)
+			if !hasConvertedPermission(rbac, binding.RoleRef.Name) {
+				// Convert role.
+				role := option.authzPolicies.RoleForNameAndNamespace(namespace, binding.RoleRef.Name)
+				for _, rule := range role.Rules {
+					// Check to make sure the services field does not exist in the ServiceRole.
+					if len(rule.Services) > 0 {
+						rbacLog.Errorf("should not have services field in ServiceRole (found in %s)", binding.RoleRef.Name)
+						continue
+					}
+					// Check to make sure that the current service (caller) is the one this rule is applying to.
+					if !service.areConstraintsMatched(rule) {
+						continue
+					}
+					if option.forTCPFilter {
+						// TODO(yangminzhu): Move the validate logic to push context and add metrics.
+						if err := validateRuleForTCPFilter(rule); err != nil {
+							// It's a user misconfiguration if a HTTP rule is specified to a TCP service.
+							// For safety consideration, we ignore the whole rule which means no access is opened to
+							// the TCP service in this case.
+							rbacLog.Debugf("rules[%d] ignored, found HTTP only rule for a TCP service: %v", i, err)
+							continue
+						}
+					}
+					// Generate the policy if the service is matched and validated to the services specified in
+					// ServiceRole.
+					permissions = append(permissions, convertToPermission(rule))
+				}
+				if len(permissions) == 0 {
+					rbacLog.Debugf("role %s skipped for no rule matched", binding.RoleRef.Name)
+					continue
+				}
+			}
+
+			if option.forTCPFilter {
+				if err := validateBindingsForTCPFilter([]*rbacproto.ServiceRoleBinding{binding}); err != nil {
+					rbacLog.Debugf("role %s skipped, found HTTP only binding for a TCP service: %v", binding.RoleRef.Name, err)
+					continue
+				}
+			}
+			// Convert binding.
+			enforcedPrincipals, _ := convertToPrincipals([]*rbacproto.ServiceRoleBinding{binding}, option.forTCPFilter)
+
+			if len(enforcedPrincipals) == 0 {
+				rbacLog.Debugf("role %s skipped for no principals found", binding.RoleRef.Name)
+				continue
+			}
+
+			policyName := "authz-policy-" + authzPolicy.Name + "-allow-" + strconv.Itoa(i)
+			rbac.Policies[policyName] = &policyproto.Policy{
+				Permissions: permissions,
+				Principals:  enforcedPrincipals,
+			}
+		}
+	}
+	return &http_config.RBAC{Rules: rbac}
+}
+
+// hasConvertedPermission returns True if there is already a |roleName| ServiceRole for |rbac|, False
+// otherwise.
+func hasConvertedPermission(rbac *policyproto.RBAC, roleName string) bool {
+	if rbac.Policies[roleName] == nil {
+		return false
+	}
+	return rbac.Policies[roleName].Permissions != nil
+}
+
+// areConstraintsMatched returns True if the calling service's attributes and/or labels match to
+// the ServiceRole constraints.
+func (service serviceMetadata) areConstraintsMatched(rule *rbacproto.AccessRule) bool {
```

Comment:
Can we put this in a single place and use it in both rbac.go and rbac_v2.go? We don't need to duplicate this as he checking logic should be the same for v1 and v2.

Reply:
Sure. I'll put it in rbac.go and used it in both places. Originally I didn't want to mess with rbac.go too much.